add encryption library service

DmyMi opened the pull request.

DmyMi committed:
add encryption library service

DmyMi committed:
upgrade user secret package

VadymLevkovskyi commented on OutOfSchool/OutOfSchool.Encryption/Config/Proxy.cs:
```diff
@@ -0,0 +1,14 @@
+
+    public string Address { get; set; }
+
+    public string Port { get; set; }
```
Is it correct ot have `Port` as a string?

h4wk13 replied:
And another question:
I haven't seen the validation anywhere... 
Could you please add the check whether the port is valid ? (>=0 && < ushort.MaxValue ) ?

DmyMi replied:
@VadymLevkovskyi @h4wk13 
Port is `string` because the `C` library methods have ports as `string` also. I can add validation, but is it necessary to do the `string -> int -> string` convertion?

VadymLevkovskyi replied:
Types are part of self-documented code. In other words, I don't expect `":8080"` value in `int` variable. Thus I'm ready to pay this "double conversion" price. 
Annotation you've used here is good as well for the mentioned purpose (though I more trust to compiler than some reflection).

VadymLevkovskyi commented on OutOfSchool/OutOfSchool.Encryption/Models/CASettings.cs:
```diff
@@ -0,0 +1,20 @@
+
+public class CASettings
+{
+    public string[] issuerCNs { get; set; }
```
(Naming) Properties are not PascalCased

VadymLevkovskyi commented on OutOfSchool/OutOfSchool.Encryption/Models/CASettings.cs:
```diff
@@ -0,0 +1,20 @@
+
+    public string ocspAccessPointAddress { get; set; }
+
+    public string ocspAccessPointPort { get; set; }
```
("Stringly-typed" code) Why port (`ocspAccessPointPort` and `tspAddressPort`) has string type?

DmyMi replied:
@VadymLevkovskyi same as above - `C` library uses it as string.

VadymLevkovskyi commented on OutOfSchool/OutOfSchool.Encryption/Models/EnvelopedUserInfoResponse.cs:
```diff
@@ -0,0 +1,10 @@
+{
+    public string EncryptedUserInfo { get; set; }
+
+    public int Error { get; set; }
```
Is this `int Error` something related to `HttpStatusCode` standard enum?

DmyMi replied:
@VadymLevkovskyi 
model is copied from 3rd party authorization API and they do return `200 OK` & error code in body. Error code is their business logic code, not exactly HTTP code. We can check for error before calling this service, will remove this here and add check later in AuthServer.

VadymLevkovskyi commented on OutOfSchool/OutOfSchool.Encryption/Handlers/AppHandlers.cs:
```diff
@@ -0,0 +1,55 @@
+
+public static class AppHandlers
+{
+    public static void Map(WebApplication app, ApiVersionSet apiVersionSet)
```
Such methods are usually implemented as extension methods (`public static WebApplication MapEUSignOAuth(this WebApplication app...`), so that they can be chained and called in similar way as standard/other 3rd party implementations.
That's not strict rule, just informaly established convention

h4wk13 commented on OutOfSchool/OutOfSchool.Encryption/Config/EUSignConfig.cs:
```diff
@@ -0,0 +1,16 @@
+
+public class EUSignConfig
+{
+    public static readonly string Name = "EUSign";
```
1. const ?
2. what the `Name` actually means here ?

VadymLevkovskyi replied:
I believe it's config section name. (Used while taking config by deserializing json/ini/other text files)

h4wk13 replied:
If so, could we please rename it to `SectionName` or `ConfigSeciontName` ?

h4wk13 commented on OutOfSchool/OutOfSchool.Encryption/Config/Proxy.cs:
```diff
@@ -0,0 +1,14 @@
+
+public class Proxy
+{
+    public bool Use { get; set; }
```
`Enabled` ?

h4wk13 commented on OutOfSchool/OutOfSchool.Encryption/Config/EUSignConfig.cs:
```diff
@@ -0,0 +1,16 @@
+
+    public CA CA { get; set; }
+
+    public Proxy Proxy { get; set; }
```
This setting here means that the system could have more than one proxy setting, is it the case ?

DmyMi replied:
@h4wk13 the configuration is parsed from `appsettings.json` and configured & injected as `IOptions` and the values are cached for the application's lifetime. Unless we implement a lot of custom logic or configure and use `IOptionsMonitor` - the only way to change these settings is to change appsettigns file or add environment variables and completely restart the application.

h4wk13 commented on OutOfSchool/OutOfSchool.Encryption/Config/Proxy.cs:
```diff
@@ -0,0 +1,14 @@
+{
+    public bool Use { get; set; }
+
+    public string Address { get; set; }
```
`Host` ?
And I haven't seen the validation if that host is valid.
 could you please add that somewhere where we load the configuration ?

h4wk13 commented on OutOfSchool/OutOfSchool.Encryption/Handlers/AppHandlers.cs:
```diff
@@ -0,0 +1,55 @@
+                "api/{version:apiVersion}/certificate",
+                ([FromServices] EUSignOAuth2Service euSignOAuth2Service) =>
+                {
+                    var cert = euSignOAuth2Service.GetEnvelopCertificateBase64();
```
GetEnvelop**e**CertificateBase64

h4wk13 commented on OutOfSchool/OutOfSchool.Encryption/Handlers/AppHandlers.cs:
```diff
@@ -0,0 +1,55 @@
+                    return Results.Problem(
+                        title: "Certificate not found",
+                        detail: "Certificate was not present in the system.",
+                        statusCode: 500);
```
`404` ?

VadymLevkovskyi replied:
I doubt - it looks more like internal server error (deployment was broken) than absence of customer's entity.

DmyMi replied:
@h4wk13 the meaning i put here is that the certificate was not present on the server or the `C` library failed to parse it. So like Vadym said - something is broken on backend.

maybe wrong wording.

h4wk13 commented on OutOfSchool/OutOfSchool.Encryption/Handlers/AppHandlers.cs:
```diff
@@ -0,0 +1,55 @@
+                        statusCode: 500);
+                })
+            .WithApiVersionSet(apiVersionSet)
+            .MapToApiVersion(1);
```
please move this and all other `1` (api version) to settings or at least to const

h4wk13 commented on OutOfSchool/OutOfSchool.Encryption/Handlers/AppHandlers.cs:
```diff
@@ -0,0 +1,55 @@
+            .MapToApiVersion(1);
+
+        app.MapPost(
+                "api/{version:apiVersion}/decryption", (
```
`decryption` > `decrypt` ?

h4wk13 commented on OutOfSchool/OutOfSchool.Encryption/Handlers/AppHandlers.cs:
```diff
@@ -0,0 +1,55 @@
+
+        app.MapPost(
+                "api/{version:apiVersion}/decryption", (
+                    EnvelopedUserInfoResponse encryptedUserInfo,
```
[NotNull] ?

DmyMi replied:
According to the [docs](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.codeanalysis.notnullattribute?view=net-8.0) it is for static analysis only. For service i agree we can add it because we control how the method is called, but for "Controller" method we have no control over how the method will be called in runtime.

h4wk13 commented on OutOfSchool/OutOfSchool.Encryption/Models/CASettings.cs:
```diff
@@ -0,0 +1,20 @@
+
+    public string address { get; set; }
+
+    public string ocspAccessPointAddress { get; set; }
```
`Address` > `Host` ?

DmyMi replied:
@h4wk13  this is parsed from 3rd party JSON. Can't rename it

h4wk13 commented on OutOfSchool/OutOfSchool.Encryption/Models/EnvelopedUserInfoResponse.cs:
```diff
@@ -0,0 +1,10 @@
+
+public class EnvelopedUserInfoResponse
+{
+    public string EncryptedUserInfo { get; set; }
```
should the model be here instead of string ?

DmyMi replied:
@h4wk13 the 3rd party API return an encrypted JSON string as a field value. We need to decrypt it and then deserialize the JSON (we can actualy skip deserialization in this service becuase we just serialize it back, but i decided to have it so we can atleast check if the JSON can bind to properties early)

h4wk13 commented on OutOfSchool/OutOfSchool.Encryption/Models/UserInfoResponse.cs:
```diff
@@ -0,0 +1,42 @@
+namespace OutOfSchool.Encryption.Models;
+
+public class UserInfoResponse
```
Pascal case for property names ?

DmyMi replied:
fixed

h4wk13 commented on OutOfSchool/OutOfSchool.Encryption/OutOfSchool.Encryption.csproj:
```diff
@@ -0,0 +1,61 @@
+      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+      <CopyToPublishDirectory>PreserveNewest</CopyToPublishDirectory>
+    </AdditionalFiles>
+    <None Remove="appsettings.jsonc" />
```
`json` ?

h4wk13 commented on OutOfSchool/OutOfSchool.Encryption/Program.cs:
```diff
@@ -0,0 +1,73 @@
+var builder = WebApplication.CreateBuilder(args);
+
+builder.Configuration
+    .AddJsonFile("appsettings.jsonc", false, true)
```
`json` ?

DmyMi replied:
@h4wk13 though `json` can have comments that are parsed by .Net, Visual Studio and Rider highlight it as errors. The `jsonc` (JSON with comments) is accepted by VS/Rider though needs some manual steps to add to project.

h4wk13 commented on OutOfSchool/OutOfSchool.Encryption/Program.cs:
```diff
@@ -0,0 +1,73 @@
+
+builder.Services.AddApiVersioning(options =>
+    {
+        options.DefaultApiVersion = new ApiVersion(1);
```
I've seen this `1` in many places.
please move this `1` to config
if this is default value - please move to the const

h4wk13 commented on OutOfSchool/OutOfSchool.Encryption/Program.cs:
```diff
@@ -0,0 +1,73 @@
+
+var app = builder.Build();
+
+app.UseSerilogRequestLogging();
```
Does that logger hide the sensitive data like certs or keys or... ?

DmyMi replied:
`UseSerilogRequestLogging()` never logs body from requests, only URLs

h4wk13 commented on OutOfSchool/OutOfSchool.Encryption/Services/DevEUSignOAuth2Service.cs:
```diff
@@ -0,0 +1,19 @@
+    /// <inheritdoc/>
+    public CertificateResponse GetEnvelopCertificateBase64() => new()
+    {
+        CertBase64 = Convert.ToBase64String("very_secret_certificate"u8.ToArray()),
```
should this name be configurable ?

DmyMi replied:
@h4wk13 dev mock authorization server has also this hardcoded as it does not implement any encryption.

h4wk13 commented on OutOfSchool/OutOfSchool.Encryption/Services/DevEUSignOAuth2Service.cs:
```diff
@@ -0,0 +1,19 @@
+    };
+
+    /// <inheritdoc/>
+    public UserInfoResponse DecryptUserInfo(EnvelopedUserInfoResponse encryptedUserInfo) =>
```
Since this is the public method could you please:
1. add the `[NotNull]` attribute to the argument
2. validate it is not null.

h4wk13 commented on OutOfSchool/OutOfSchool.Encryption/Services/EUSignOAuth2Exception.cs:
```diff
@@ -0,0 +1,21 @@
+
+public class EUSignOAuth2Exception : Exception
+{
+    public EUSignOAuth2Exception()
```
?

h4wk13 commented on OutOfSchool/OutOfSchool.Encryption/Services/IEUSignOAuth2Service.cs:
```diff
@@ -0,0 +1,24 @@
+
+namespace OutOfSchool.Encryption.Services;
+
+public interface IEUSignOAuth2Service
```
I suppose we can add the documentation for interface as well )
And: we might use the `/// <inheritdoc/>` in all the implementations for each method so they will be documented.

h4wk13 commented on OutOfSchool/OutOfSchool.Encryption/stylecop.json:
```diff
@@ -0,0 +1,14 @@
+{
+  // ACTION REQUIRED: This file was automatically added to your project, but it
```
?

VadymLevkovskyi commented on OutOfSchool/OutOfSchool.Encryption/Services/DevEUSignOAuth2Service.cs:
```diff
@@ -0,0 +1,19 @@
+    /// <inheritdoc/>
+    public CertificateResponse GetEnvelopCertificateBase64() => new()
+    {
+        CertBase64 = Convert.ToBase64String("very_secret_certificate"u8.ToArray()),
```
(Security) `very_secret_certificate` )))

VadymLevkovskyi replied:
ok, let it be, though it could be at least some constant with meaningful name

VadymLevkovskyi commented on OutOfSchool/OutOfSchool.Encryption/Services/EUSignOAuth2Service.cs:
```diff
@@ -0,0 +1,480 @@
+                out var developedUserInfo,
+                out _);
+
+            Stream userInfoStream = new MemoryStream(developedUserInfo);
```
(Resource deallocation) Please use `using` here

VadymLevkovskyi commented on OutOfSchool/OutOfSchool.Encryption/Services/EUSignOAuth2Service.cs:
```diff
@@ -0,0 +1,480 @@
+    /// Пошук зареєстрованого носія ключової інформації
+    /// This code is from IIT Library usage example with minimal code styling changes.
+    /// </summary>
+    private static void GetKeyMedia(
```
(Codestyle) This whole method is written using non-C# code style

VadymLevkovskyi replied:
Ok, seems this refactoring could be postponed

VadymLevkovskyi commented on OutOfSchool/OutOfSchool.Encryption/Services/EUSignOAuth2Service.cs:
```diff
@@ -0,0 +1,480 @@
+            string curType, curDevice;
+
+            typeIndex = 0;
+            while (true)
```
That's very brave idea to not limit `typeIndex` and `deviceIndex` in these `while(true)` cycles

h4wk13 replied:
You've probably meant that: if device name from the config will be wrong - we will have to wait for some noticeable time while got the very descriptive error.

VadymLevkovskyi replied:
> By default, integral values (such as int, uint, and long) do not throw an exception when they overflow. Instead, they "wrap" -- and this is probably not a value that we want. But we can change this behavior so that an exception is thrown instead. [Link](
https://jeremybytes.blogspot.com/2023/01/checking-for-overflow-in-c.html#:~:text=January%2017%2C%202023-,Checking%20for%20Overflow%20in%20C%23,an%20exception%20is%20thrown%20instead.)

"Noticable" time could last forever.

VadymLevkovskyi commented on OutOfSchool/OutOfSchool.Encryption/Services/EUSignOAuth2Service.cs:
```diff
@@ -0,0 +1,480 @@
+        // Необхідні при обслуговуванні користувачів з різних ЦСК
+        IEUSignCP.SetOCSPAccessInfoModeSettings(true);
+
+        var fsCAs = new FileStream(eUSignConfig.CA.JsonPath, FileMode.OpenOrCreate);
```
(Resources deallocation) `Using`

VadymLevkovskyi commented on OutOfSchool/OutOfSchool.Encryption/appsettings.jsonc:
```diff
@@ -0,0 +1,96 @@
+      "MediaDevice": "001:3011 (192.168.0.100)",
+      // Password of the server's private key.
+      // If Гряда-301 is used as the medium, the username and password are set as ##Login##Password
+      "Password": "12345678",
```
(Security) Though it looks like a test password it's not a good idea to have it in config

h4wk13 commented on OutOfSchool/OutOfSchool.Encryption/Services/EUSignOAuth2Service.cs:
```diff
@@ -0,0 +1,480 @@
+                throw new EUSignOAuth2Exception("An error occurred while initializing the cryptographic library.", ex);
+            }
+
+            ServicePointManager.SecurityProtocol = SecurityProtocolType.Tls12;
```
I believe it is the default setting for current .NET version.
Should we also add 1.3 version ?

h4wk13 commented on OutOfSchool/OutOfSchool.Encryption/Services/EUSignOAuth2Service.cs:
```diff
@@ -0,0 +1,480 @@
+    /// <inheritdoc/>
+    public CertificateResponse GetEnvelopCertificateBase64()
+    {
+        if (pkEnvelopCert == null)
```
pkEnvelop**e**Cert

h4wk13 commented on OutOfSchool/OutOfSchool.Encryption/Services/EUSignOAuth2Service.cs:
```diff
@@ -0,0 +1,480 @@
+    {
+        if (pkEnvelopCert == null)
+        {
+            logger.LogError("No envelop certificate was loaded");
```
envelop**e**

h4wk13 commented on OutOfSchool/OutOfSchool.Encryption/Services/EUSignOAuth2Service.cs:
```diff
@@ -0,0 +1,480 @@
+    }
+
+    /// <inheritdoc/>
+    public CertificateResponse GetEnvelopCertificateBase64()
```
Do we really need to return the cert instead of encrypt something ?

DmyMi replied:
@h4wk13 the certificate is used by Authorization server to send UserInfo request during the OAuth2 flow. So we call this service, get the certificate and send it to 3rd party OAuth provider so it encrypts data.

h4wk13 commented on OutOfSchool/OutOfSchool.Encryption/Services/EUSignOAuth2Service.cs:
```diff
@@ -0,0 +1,480 @@
+    }
+
+    /// <inheritdoc/>
+    public UserInfoResponse DecryptUserInfo(EnvelopedUserInfoResponse encryptedUserInfo)
```
[NotNull]

h4wk13 commented on OutOfSchool/OutOfSchool.Encryption/Services/EUSignOAuth2Service.cs:
```diff
@@ -0,0 +1,480 @@
+            IEUSignCP.EU_SAVE_SETTINGS_PARAMETER, IEUSignCP.EU_SETTINGS_ID_NONE);
+
+        // Вимкнення відображення помилок
+        IEUSignCP.SetUIMode(false);
```
it was called few lines above

DmyMi replied:
@h4wk13 it is done in the examples in all languages. Probably some logic between those calls resets this value :)

h4wk13 replied:
or it is a mistake in example, which is highly possible I suppose )

h4wk13 commented on OutOfSchool/OutOfSchool.Encryption/Services/EUSignOAuth2Service.cs:
```diff
@@ -0,0 +1,480 @@
+        IEUSignCP.EU_KEY_MEDIA keyMedia = new IEUSignCP.EU_KEY_MEDIA(0, 0, string.Empty);
+        CASettings ca = null;
+        IEUSignCP.EU_CERT_OWNER_INFO info;
+        IEUSignCP.EU_CERT_INFO_EX infoEx;
```
could you please review these variables and move declarations to the places where they really used ?

h4wk13 commented on OutOfSchool/OutOfSchool.Encryption/Services/EUSignOAuth2Service.cs:
```diff
@@ -0,0 +1,480 @@
+                        throw new Exception($"The certificate file is missing: {file}");
+                    }
+
+                    certs.Add(File.ReadAllBytes(file));
```
How many of certs we can load into memory on the server without experiencing memory issues ?

DmyMi replied:
@h4wk13  according to documentation there can be 1-2 certificate files, as we are going to host only 1 organization

h4wk13 commented on OutOfSchool/OutOfSchool.Encryption/Services/EUSignOAuth2Service.cs:
```diff
@@ -0,0 +1,480 @@
+                if (!string.IsNullOrEmpty(jksAlias))
+                {
+                    // Пошук відповідного ос. ключа в контейнері JKS
+                    IEUSignCP.GetJKSPrivateKey(pKey, jksAlias, out pKey, out var certificates);
```
it seems that we're going to overwrite `pKey` variable initialized in the 340 line. Is it OK ? if yes - why ?

DmyMi replied:
@h4wk13 according to docs certain certificate authorities have certificates as a container for set of different keys instead of being plain certificate. So we read the certificate and if it is a container (we specify the alias to search) - we use it and the key alias to find the actual certificate in the container

h4wk13 replied:
Sounds promising...

h4wk13 commented on OutOfSchool/OutOfSchool.Encryption/Services/EUSignOAuth2Service.cs:
```diff
@@ -0,0 +1,480 @@
+    private IntPtr context = IntPtr.Zero;
+    private IntPtr pkContext = IntPtr.Zero;
+    private byte[] pkSignCert = null;
+    private byte[] pkEnvelopCert = null;
```
could you please add `readonly` modifier too all of these fields that are initialized in scope or constructor ?

DmyMi replied:
@h4wk13 it is initialized in constructor, but as an `out` property

> Field 'pkEnvelopeCert' is read-only (except in constructor or init-only setter of the type in which the field is defined). 'out' argument must be an assignable variable, field, or an array element

h4wk13 commented on OutOfSchool/OutOfSchool.Encryption/Services/EUSignOAuth2Service.cs:
```diff
@@ -0,0 +1,480 @@
+        // - з параметрів TSP, що встановлені за замовчанням.
+
+        // Встановлення параметрів TSP-серверу ЦСК за замовчанням
+        IEUSignCP.SetTSPSettings(true, eUSignConfig.DefaultTSPServer, "80");
```
can we read that and all other `"80"` from config ?

h4wk13 commented on OutOfSchool/OutOfSchool.Encryption/Services/EUSignOAuth2Service.cs:
```diff
@@ -0,0 +1,480 @@
+            if (fromFile)
+            {
+                // Зчитування ос. ключа з файлу до байтового масиву
+                if (!File.Exists(pkFile))
```
I suppose we would like to have unit tests for that service, if so - could you please move all System.IO related calls into dedicated helper or service or... and inject it as the constructor argument ?

h4wk13 commented on OutOfSchool/OutOfSchool.Encryption/Services/EUSignOAuth2Service.cs:
```diff
@@ -0,0 +1,480 @@
+                {
+                    // Пошук відповідного ос. ключа в контейнері JKS
+                    IEUSignCP.GetJKSPrivateKey(pKey, jksAlias, out pKey, out var certificates);
+                    foreach (var certificate in certificates)
```
please replace it with certs.AddRange()

h4wk13 commented on OutOfSchool/OutOfSchool.Encryption/Services/EUSignOAuth2Service.cs:
```diff
@@ -0,0 +1,480 @@
+            }
+            catch (Exception)
+            {
+                throw new Exception("The private key does not support PKI encryption described in ДСТУ 4145");
```
We've lost the original exception here, could we log it with the error ? 
If it is taken as is from the example - it might be the risk to log due to sensitive data exposure...

DmyMi replied:
@h4wk13 according to their source code the exception text is generated by `C` code. No indication if it is sensitive or not.

DmyMi committed:
fixed comments

add port validation

remove error from DTO as it should be validated in AuthServer

added System.Io wrapper

fix constant usage

fix naming and types

DmyMi committed:
moved doc to wiki

DmyMi committed:
add cloud build config

DmyMi committed:
add a limit to loop

DmyMi committed:
add check to see if EUSign is initialized

DmyMi committed:
removed stylecop from service

h4wk13 commented on OutOfSchool/OutOfSchool.Encryption/Config/EUSignConfig.cs:
```diff
@@ -0,0 +1,24 @@
+{
+    public static readonly string ConfigSectionName = "EUSign";
+
+    public string DefaultOCSPServer { get; set; }
```
could/should we add `[Required]` here and for other ref types ?

h4wk13 commented on OutOfSchool/OutOfSchool.Encryption/Services/EUSignOAuth2Service.cs:
```diff
@@ -0,0 +1,485 @@
+    private void Initialize()
+    {
+        // Увімкнення кидання винятків бібліотекою, замість повертання помилок.
+        IEUSignCP.SetThrowExceptions(true);
```
should these flags be moved to config ?

h4wk13 commented on OutOfSchool/OutOfSchool.Encryption/Services/EUSignOAuth2Service.cs:
```diff
@@ -0,0 +1,485 @@
+        // Встановлення налаштувань Сховища сертифікатів та СВС
+        IEUSignCP.SetFileStoreSettings(
+            string.Empty,
+            false,
```
configuration ?
I would ask you to move all these `1,2,3, true, false` and other literals to the config or to the constants, if you don't mind.

h4wk13 commented on OutOfSchool/OutOfSchool.Encryption/Services/EUSignOAuth2Service.cs:
```diff
@@ -0,0 +1,485 @@
+                        keyInfo, cmpServers, cmpServersPorts, out certsCMP);
+
+                    int i = 0;
+                    while (true)
```
should it also be changed to `i < int.MaxValue ?` or to `for`

DmyMi committed:
fix comments

DmyMi committed:
extracted more options to config
